tools/runqslower.py

```diff
@@ -167,19 +174,26 @@
     // TP_PROTO(bool preempt, struct task_struct *prev, struct task_struct *next)
     struct task_struct *prev = (struct task_struct *)ctx->args[1];
     struct task_struct *next= (struct task_struct *)ctx->args[2];
-    u32 pid;
+    u32 tgid, pid;
     long state;
 
     // ivcsw: treat like an enqueue event and store timestamp
     bpf_probe_read(&state, sizeof(long), (const void *)&prev->state);
     if (state == TASK_RUNNING) {
+        bpf_probe_read(&tgid, sizeof(prev->tgid), &prev->tgid);
         bpf_probe_read(&pid, sizeof(prev->pid), &prev->pid);
-        if (!(FILTER_PID || pid == 0)) {
-            u64 ts = bpf_ktime_get_ns();
-            start.update(&pid, &ts);
+
+        u64 ts = bpf_ktime_get_ns();
+        if (pid != 0) {
+            if ( (!(FILTER_PID) && FILTER_TGID == 0) ||
+                 (!(FILTER_TGID) && FILTER_PID == 0)) {
+                    start.update(&pid, &ts);
+            }
         }
+
     }
 
+    bpf_probe_read(&tgid, sizeof(next->tgid), &next->tgid);
```

Comment:
This probably not needed. Could you double check? We have another place to bpf_probe_read tgid as well, you may want to check as well.